### PR TITLE
add another wamr test

### PR DESCRIPTION
**Pull request by Zzzabiyaka**
follows up #2364 where we discussed that we might want to have a test which has really short thread function and creates many threads

### CODE REVIEW

File: .github/workflows/compilation_on_android_ubuntu.yml
Comment (loganek) [inline .github/workflows/compilation_on_android_ubuntu.yml:530]: How long does it take to run the stress test? If that all adds up to 60 minutes now, I think it's reasonable to keep for nightly only.

File: core/iwasm/libraries/lib-wasi-threads/test/spawn_stress_test.c
Comment (loganek) [inline core/iwasm/libraries/lib-wasi-threads/test/spawn_stress_test.c:23]: worth adding a unit to the name, so the reader doesn't have to look for the usage to know the meaning of the number.
```suggestion
    RETRY_SLEEP_TIME_US = 8000,
```

File: core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c
Comment (loganek) [inline core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c:7]: not needed I guess?

File: core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c
Comment (loganek) [inline core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c:43]: I wonder if exponential backoff wouldn't be better here; if we always assume 8ms sleep time (which is quite a lot IMO for this type of test), we might not be stressing the implementation enough.

File: core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c
Comment (loganek) [inline core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c:64]: For this test I think we should spawn threads as quickly as possible - here we wait for a specific thread ID to be released, even though other IDs might have been freed already and could be used. I think this was discussed for the other test too, but I don't know if that was eventually addressed? If so, the code probably could be reused.

File: core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c
Comment (loganek) [inline core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c:80]: where does the `1.05` number come from? Have you tested it on different configurations? What if we run the test on some low-end devices - would the test fail?

File: core/iwasm/libraries/lib-wasi-threads/test/skip.json
Comment (loganek) [inline core/iwasm/libraries/lib-wasi-threads/test/skip.json:4]: more tests we add here more convinced I am this should all be ran as part of the test runner. What would it take to run those tests just like we run other thread tests?

File: .github/workflows/compilation_on_android_ubuntu.yml
Comment (Zzzabiyaka) [reply .github/workflows/compilation_on_android_ubuntu.yml:530]: it's a bit tricky 

actually it should be 2 times less. This constraint applies because `jit` is basically 2 bundle runs 

here's the ticket to set timeouts based on the platform. So, all platforms <= 30 minutes, jit <= 60 minutes

File: core/iwasm/libraries/lib-wasi-threads/test/skip.json
Comment (Zzzabiyaka) [reply core/iwasm/libraries/lib-wasi-threads/test/skip.json:4]: for AOT - nothing 

for non-aot I don't even know how hard will it be (because aot is executed differently)

File: core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c
Comment (Zzzabiyaka) [reply core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c:80]: No I didn't 

1.05 is meant to be "a bit higher than 1", e.g. to allow some retries but not that much relatively to created threads. 
 
Maybe I need to double-check on some low devices

File: core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c
Comment (Zzzabiyaka) [reply core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c:64]: I'd expect that we use this part of pthread_join

```  
If that thread has already terminated, then
       pthread_join() returns immediately. 
```

I'm curious whether there's any better solution. If we have mutex, thread function is going to be much longer I suppose. 

If we have atomic, we most likely need SEQ order, also not the fastest solution. 

I think that `thread_execution_time` is going to be less than `7 * thread_creation_time` for this tiny thread function. 

But it's just my assumation, tell me if I'm wrong

File: core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c
Comment (Zzzabiyaka) [reply core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c:43]: hm, interesting idea, I will think about it. Thanks!

Comment (Zzzabiyaka) [reply core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c:43]: less than 8ms was introducing some flakiness where test could randomly fail

File: core/iwasm/libraries/lib-wasi-threads/test/spawn_stress_test.c
Comment (Zzzabiyaka) [reply core/iwasm/libraries/lib-wasi-threads/test/spawn_stress_test.c:23]: Thanks will do

File: .github/workflows/compilation_on_android_ubuntu.yml
Comment (loganek) [reply .github/workflows/compilation_on_android_ubuntu.yml:530]: Even if

File: core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c
Comment (loganek) [reply core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c:64]: So imagine we have three threads (call it 1, 2 and 3). Potential sequence can be:
```
Spawn thread 1
Spawn thread 2
Spawn thread 3
Finish thread 2
Finish thread 3
Spawn thread 1 <- block here
```
Because the first thread is still alive (e.g. due to scheduling), we can't spawn new thread even though slot 2 and 3 are freed already, so we wait unnecessarily.
Not sure how likely or severe the issue is, but it feels like we can completely remove the problem with a low overhead by not relying on the order.

File: .github/workflows/compilation_on_android_ubuntu.yml
Comment (wenyongh) [reply .github/workflows/compilation_on_android_ubuntu.yml:530]: Seems it took long time to run the stress cases:
https://github.com/bytecodealliance/wasm-micro-runtime/actions/runs/5761023933/job/15618324568
https://github.com/bytecodealliance/wasm-micro-runtime/actions/runs/5761023933/job/15618325167
https://github.com/bytecodealliance/wasm-micro-runtime/actions/runs/5761023933/job/15618325916
https://github.com/bytecodealliance/wasm-micro-runtime/actions/runs/5761023933/job/15618326547
https://github.com/bytecodealliance/wasm-micro-runtime/actions/runs/5761023933/job/15618326990

Almost close to 30 min for each running mode, but it is only 4-5 min now:
https://github.com/bytecodealliance/wasm-micro-runtime/actions/runs/5791716284/job/15697007728
https://github.com/bytecodealliance/wasm-micro-runtime/actions/runs/5791716284/job/15697008239
https://github.com/bytecodealliance/wasm-micro-runtime/actions/runs/5791716284/job/15697009623

Shall we put the stress test in nightly run?

Comment (Zzzabiyaka) [reply .github/workflows/compilation_on_android_ubuntu.yml:530]: we can either decrease the constants (what should be harmless, because if it did 100'000 threads it's really likely to do 10 times more without errors or put to nightly

Comment (wenyongh) [reply .github/workflows/compilation_on_android_ubuntu.yml:530]: How about decreasing the constants to ensure it doesn't increase too much time in this PR, and then adding another PR for stress test?
We had better not make it take too long to run the CIs of a PR, developer may be eager to see the running result.

File: core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c
Comment (Zzzabiyaka) [reply core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c:64]: Fixed

File: core/iwasm/libraries/lib-wasi-threads/test/skip.json
Comment (Zzzabiyaka) [reply core/iwasm/libraries/lib-wasi-threads/test/skip.json:4]: Maybe we can go the opposite way and scope threads that depend on max_threads=4 (there should be only one/a couple of them)

File: .github/workflows/compilation_on_android_ubuntu.yml
Comment (Zzzabiyaka) [reply .github/workflows/compilation_on_android_ubuntu.yml:530]: I've changed the constants. I will wait for execution and will adjust timeouts respectively

Comment (wenyongh) [reply .github/workflows/compilation_on_android_ubuntu.yml:530]: OK, the nightly_run CI reported failure:

https://github.com/bytecodealliance/wasm-micro-runtime/actions/runs/5801072533/job/15724690363#step:12:488

```
stress_test_threads_creation.wasm is a stress test
Assertion failed: status_code == 0 && "Thread creation should succeed" (stress_test_threads_creation.c: spawn_thread: 55)
Exception: unreachable
Stress test stress_test_threads_creation.wasm FAILED with code  1
```

Comment (Zzzabiyaka) [reply .github/workflows/compilation_on_android_ubuntu.yml:530]: for some reason `multitier-jit` looks really problematic in terms of threads creation. 

I'm not sure whether it's by design or it can be a potential sign of some bug

File: core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c
Comment (loganek) [inline core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c:67]: Since you detach the thread, you probably don't need this array any more?

File: core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c
Comment (loganek) [inline core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c:23]:
```suggestion
pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
```

and then you can remove init/destroy calls in the main function.

File: core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c
Comment (loganek) [inline core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c:33]: won't that line be removed by compiler's optimizer?

File: core/iwasm/libraries/lib-wasi-threads/test/build.sh
Comment (loganek) [inline core/iwasm/libraries/lib-wasi-threads/test/build.sh:13]: for this new test we probably want to enable compiler optimizations so we can spawn as many threads as we can.

File: core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c
Comment (loganek) [inline core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c:88]: do we need to know which thread finished? Wouldn't be enough to know how many threads is running?

**Review by loganek — CHANGES_REQUESTED** [review]

File: core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c
Comment (Zzzabiyaka) [reply core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c:88]: I want to reuse `pthread_t` array so I need to know which of `threads[i]` I need to rewrite

File: core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c
Comment (Zzzabiyaka) [reply core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c:33]: I don't think so

I mean, num var is used in the next line so you can't remove it's initialising, if I got your question right

File: core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c
Comment (Zzzabiyaka) [reply core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c:67]: hm, 

I had a concern that `pthread_t` var assigned to thread should be alive until thread finishes. 

I'm not a posix expert so I'm not 100% so tell me if I'm wrong

File: core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c
Comment (Zzzabiyaka) [reply core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c:23]: thanks

File: core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c
Comment (Zzzabiyaka) [reply core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c:33]: I don't think so

I mean, num var is used in the next line so you can't remove it's initialising, if I got your question right

File: core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c
Comment (Zzzabiyaka) [reply core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c:66]: hm,

I had a concern that pthread_t var assigned to thread should be alive until thread finishes.

I'm not a posix expert so I'm not 100% sure so tell me if I'm wrong

File: core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c
Comment (loganek) [reply core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c:33]: Ah yes, sorry I didn't provide a full context in this comment. I also assumed we don't need the is_finished array

File: core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c
Comment (loganek) [reply core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c:66]: You don't need to keep track on the thread I'd if you don't use it. `pthread_t` in wasi libc (and many other implementations) is simply an integer. Also, C doesn't have RAII by default soeven if it was a complex data structure allowing thread management, worse case you'd have a memory leak, but thread wouldn't be affected unless some function was explicitly called

File: core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c
Comment (loganek) [reply core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c:87]: As said in the comment above, you don't need to track it

File: core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c
Comment (Zzzabiyaka) [reply core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c:66]: thanks. Got it

File: tests/wamr-test-suites/wasi-test-script/run_wasi_tests.sh
Comment (wenyongh) [inline tests/wamr-test-suites/wasi-test-script/run_wasi_tests.sh:17]: Must we add `-v=5`? It may print a lot of logs.

**Review by wenyongh — APPROVED** [review]
LGTM

@Zzzabiyaka @loganek Is it OK merge this PR now?

File: core/iwasm/libraries/lib-wasi-threads/test/skip.json
Comment (Zzzabiyaka) [reply core/iwasm/libraries/lib-wasi-threads/test/skip.json:4]: let's do this in another PR

File: core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c
Comment (Zzzabiyaka) [reply core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c]: now it's 2.5 

i'll merge a fix if it's not right for low-end devices after testing. I think we shouldn't block this PR on that

File: core/iwasm/libraries/lib-wasi-threads/test/build.sh
Comment (Zzzabiyaka) [reply core/iwasm/libraries/lib-wasi-threads/test/build.sh:37]: that makes sense and I think we can safely set -O2 for all the tests here. Does it make sense to you?

File: core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c
Comment (loganek) [inline core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c:68]: did you consider using conditionals or atomic waits here instead of sleep?

File: core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c
Comment (loganek) [inline core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c:76]: should we wait here for all the threads to complete?

File: tests/wamr-test-suites/wasi-test-script/run_wasi_tests.sh
Comment (loganek) [inline tests/wamr-test-suites/wasi-test-script/run_wasi_tests.sh:72]: when is this condition not true?

File: core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c
Comment (Zzzabiyaka) [reply core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c:76]: we can't probably wait for them, as they're detached, but at least we can wait until threads_in_use is zero

Comment (loganek) [reply core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c:76]: yes, I meant waiting for `threads_in_use` to be zero

File: core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c
Comment (Zzzabiyaka) [reply core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c:68]: we can do that I think (especially atomic wait) but I'm not sure what the benefit will we get comparing to just busy wait

File: tests/wamr-test-suites/wasi-test-script/run_wasi_tests.sh
Comment (Zzzabiyaka) [reply tests/wamr-test-suites/wasi-test-script/run_wasi_tests.sh:72]: We're just not executing the test as the wasm doesn't exist. 

we can just move `echo` inside the if

Comment (Zzzabiyaka) [reply tests/wamr-test-suites/wasi-test-script/run_wasi_tests.sh:71]: if you don't want to run stress tests for any reasons , the fastest way for you is just to delete .wasm files.
You don't have any other easy ways to avoid running it, so until you're able to disable them explicitly (e.g. until test-runner and refactoring test running system in wamr ) I'd probably leave an opportunity to avoid running stress tests. 

But generally I don't mind and can delete this line

File: core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c
Comment (Zzzabiyaka) [reply core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c:77]: done, thanks

File: core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c
Comment (Zzzabiyaka) [reply core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c:69]: and how can we do that actually? 
we can introduce a cond var that will be notifying when thread finishes but I think it's just making the test a bit more complex without any benefit (and most likely slows it as well because I don't think it will be any faster using cond var

Comment (loganek) [reply core/iwasm/libraries/lib-wasi-threads/test/stress_test_threads_creation.c:69]: the advantage is that you won't have to wait for 0.1ms for the thread to complete if it completes sooner than that. But it's up to you whether to keep it like that or not, I don't insist to change it.

**Review by loganek — APPROVED** [review]

**Pull request merged**
